### PR TITLE
Account for monotonically increasing offsets across multiple batches on arrow-CH conversion

### PR DESCRIPTION
**Pull request by arthurpassos**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
This is actually a bug fix, but not sure I'll be able to add a test to support the case, so I have put it as an improvement. This issue was introduced in https://github.com/ClickHouse/ClickHouse/pull/45878, which is when CH started reading arrow in batches.

CH uses element size as "offsets", while arrow uses actual offsets as offsets. That's why CH usually starts reading offsets with i=1 and i=0 is ignored. In case multiple batches are used to read a column, there is a chance the offsets are
monotonically increasing, which will cause inconsistencies with the batch data length on `DB::ColumnArray`.

There are two cases:
1. If the offsets are monotonically increasing, `arrow_offsets.Value(0)` will be non-zero for the nth batch, where n > 0.
2. If they are not monotonically increasing, it'll always be 0.

Therefore, we subtract the previous offset from the current offset to get the corresponding CH "offset".

This workaround should not be confused with the one implemented in https://github.com/ClickHouse/ClickHouse/pull/43297. That workaround deals with multiple chunks and internal representation of offsets, not batches.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit ce673d6beab8203b95b50feea8862038ab2bee51 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/54370/ce673d6beab8203b95b50feea8862038ab2bee51/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

File: src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp [hunk 1]
```diff
@@ -447,10 +447,25 @@ static ColumnPtr readOffsetsFromArrowListColumn(std::shared_ptr<arrow::ChunkedAr
...
+         * This workaround should not be confused with the one above. That workaround deals with multiple chunks and internal representation
+         * of offsets, not batches.
+         * */
+        auto previous_offset = arrow_offsets.Value(0);
+
         for (int64_t i = 1; i < arrow_offsets.length(); ++i)
         {
             auto offset = arrow_offsets.Value(i);
-            offsets_data.emplace_back(start_offset + offset);
+            offsets_data.emplace_back(start_offset + offset - previous_offset);
```
[the remark this answers is not in the document]
Comment (arthurpassos) [reply src/Processors/Formats/Impl/ArrowColumnToCHColumn.cpp]: That does make sense, updated the code

**Comment by alexey-milovidov** [comment]
Test needed.

**Review by Avogar — APPROVED** [review]

**Comment by arthurpassos** [comment]
Generated a file that reproduces the issue with the below code:

```
std::string random_string(size_t length) {
  static const std::string characters = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

  std::random_device random_device;
  std::mt19937 generator(random_device());
  std::uniform_int_distribution<> distribution(0, characters.size() - 1);

  std::string random_string;
  random_string.reserve(length);

  std::generate_n(std::back_inserter(random_string), length, [&]() {
    return characters[distribution(generator)];
  });

  return random_string;
}

static const std::string the_string = random_string(9247124u);

std::shared_ptr<arrow::Array> CreateIntArray(std::size_t length) {
  arrow::MemoryPool* pool = arrow::default_memory_pool();

  auto int_builder_ptr = std::make_shared<arrow::Int64Builder>(pool);
  auto & int_builder = *int_builder_ptr;
  arrow::ListBuilder list_builder(pool, int_builder_ptr);

  for (auto i = 0u; i < length; i++)
  {
    if (i % 10 == 0)
    {
      ARROW_CHECK_OK(list_builder.Append());
    }
    else
    {
      ARROW_CHECK_OK(int_builder.Append(i));
    }
  }

  std::shared_ptr<arrow::Array> int_list_array;
  ARROW_CHECK_OK(list_builder.Finish(&int_list_array));
  return int_list_array;
}

std::shared_ptr<arrow::Array> CreateStringArray(std::size_t length) {
  arrow::MemoryPool* pool = arrow::default_memory_pool();

  auto str_builder = std::make_shared<arrow::LargeStringBuilder>(arrow::large_utf8(), pool);

  for (auto i = 0u; i < length; i++)
  {
    if (i % 10 == 0)
    {
      ARROW_CHECK_OK(str_builder->AppendNull());
    }
    else
    {
      ARROW_CHECK_OK(str_builder->Append(the_string));
    }
  }

  std::shared_ptr<arrow::Array> str_array;
  ARROW_CHECK_OK(str_builder->Finish(&str_array));
  return str_array;
}

void run()
{
  auto schema = arrow::schema({
      arrow::field("ints", arrow::list(arrow::int64())),
      arrow::field("strings", arrow::utf8())
  });

  auto l1_length = 2000;
  auto l2_length = 2000;

  std::vector<std::shared_ptr<arrow::RecordBatch>> batches;

  auto int_array1 = CreateIntArray(l1_length);

  auto int_array2 = CreateIntArray(l1_length);

  auto str_array1 = CreateStringArray(l2_length);

  auto str_array2 = CreateStringArray(l2_length);

  batches.push_back(arrow::RecordBatch::Make(schema, int_array1->length(), {int_array1, str_array1}));

  batches.push_back(arrow::RecordBatch::Make(schema, int_array2->length(), {int_array2, str_array2}));

  std::shared_ptr<arrow::io::FileOutputStream> outfile;
  PARQUET_ASSIGN_OR_THROW(outfile, arrow::io::FileOutputStream::Open("generated.parquet"));

  parquet::WriterProperties::Builder builder;
  builder.compression(arrow::Compression::GZIP);
  builder.dictionary_pagesize_limit(10*1024*1024); // Given your string size is roughly 9.2MB
  builder.data_pagesize(20*1024*1024); // Given your dataset's repetition, let's try with 20MB

  std::shared_ptr<parquet::WriterProperties> props = builder.build();

  std::unique_ptr<parquet::arrow::FileWriter> file_writer;
  PARQUET_ASSIGN_OR_THROW(file_writer, parquet::arrow::FileWriter::Open(*schema, ::arrow::default_memory_pool(), outfile, props));

  for (const auto& batch : batches) {
    PARQUET_THROW_NOT_OK(file_writer->WriteRecordBatch(*batch));
  }

  PARQUET_THROW_NOT_OK(file_writer->Close());
}
```

**Comment by Avogar** [comment]
ASTFuzzer: https://github.com/ClickHouse/ClickHouse/issues/54606
Upgrade check (tsan) - known issue
FunctionalStatelessTestFlakyCheck - github action failed due to timeout, so it's ok

File: tests/queries/0_stateless/02874_parquet_multiple_batches_array_inconsistent_offsets.sh [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+#  }
+#
+#  PARQUET_THROW_NOT_OK(file_writer->Close());
+#}
+
+DATA_FILE=$CUR_DIR/data_parquet/string_int_list_inconsistent_offset_multiple_batches.parquet
+${CLICKHOUSE_CLIENT} --query="DROP TABLE IF EXISTS parquet_load"
+${CLICKHOUSE_CLIENT} --query="CREATE TABLE parquet_load (ints Array(Int64), strings Nullable(String)) ENGINE = Memory"
+cat "$DATA_FILE" | ${CLICKHOUSE_CLIENT} -q "INSERT INTO parquet_load FORMAT Parquet"
+${CLICKHOUSE_CLIENT} --query="SELECT * FROM parquet_load" | md5sum
```
Comment (den-crane) [inline tests/queries/0_stateless/02874_parquet_multiple_batches_array_inconsistent_offsets.sh:127]: >"SELECT * FROM parquet_load" | md5sum

there is no guarantee that rows will be always returned in the same stable order for engine=Memory. I would try to stabilize it using `order by strings, ints`

**Pull request merged**